### ISSUE TITLE
Make rules work with Python's Flask framework

### ISSUE THREAD

**Issue by fmkone** [jruizgit/rules#47]
Project seems exciting. 
I created a test project and tried to integrate rules with Flask's end points. If i put flask code on top then i get json response what Flask sends. It does not trigger rule which is mentioned.
My code is below:

`from durable.lang import *
from flask import Flask
app = Flask(**name**)
@app.route('/test/1', methods=['POST'])
def test():
    return "Hello, cool!"

if **name** == '**main**':
    app.run(debug=True)
with ruleset('test'):
    # antecedent
    @when_all(m.subject == 'World')
    def say_hello(c):
        # consequent
        print('Hello World {0}'.format(c.m.subject))

```
@when_start
def start(host):
    host.post('test', {'id': 1, 'sid': 1, 'subject': 'World'})
```

run_all()

`

If i put flask's route on top of rules then i get json response which is defined. But if i move rules cod on top then i get this response:
{"outcome": 212}

and no Flask's json.

How to integrate it with Flask?

**Comment by jruizgit** [comment]
Hi, thanks a lot for posting the question. durable_rules creates a Werkzeug app when run_all is called. Unfortunately there is no easy way to aggregate such an app with the Flask app. You can always create the hosts and dispatch the events from you Flask app. For example:

```python
from durable.lang import *
from flask import Flask, request, make_response
import json

with ruleset('test'):
    # antecedent
    @when_all(m.subject == 'World')
    def say_hello(c):
        # consequent
        print('Hello World {0}'.format(c.m.subject))

    @when_start
    def start(host):
        host.post('test', {'id': 1, 'sid': 1, 'subject': 'World'})

host = create_host()

flaskApp = Flask('name')
@flaskApp.route('/test', methods=['GET'])
def test():
    return "Hello, cool!"

@flaskApp.route('/test/<sid>', methods=['POST'])
def post_event():
    message = json.loads(request.stream.read())
    message['sid'] = sid
    result = host.post('test', message)
    return make_response(json.dumps({'outcome': result}))
    
flaskApp.run()
```

**Comment by fmkone** [comment]
Thanks for the response. 
We changed our plan a bit and now trying to integrate it with Node and Express server. 
And we run into the same problem. 
It seems express and node do not respond to rules at all. Rest of the express app is working but rules just don't respond or show anything.
Could you share a small snippet with Node & Express please?

**Comment by fmkone** [comment]
Here is the code i am using:

`var express = require('express');
var durable = require('durable');
var app = express();


var d = durable;
var m = d.m, s = d.s, c = d.c;
var timeout = d.timeout;
durable.ruleset('test', {
        whenAll: [
            c.first = m.location.eq('US'),
            c.second = m.location.eq('CA')
        ],
        run: function(c) {
            console.log('test fraud  detected ' + c.first.amount + ' ' + c.second.amount);
        }
    },
    function (host) {
        host.post('test', {id: 1, sid: 1, amount: 200, location: 'US'});
        host.post('test', {id: 2, sid: 1, amount: 200, location: 'CA'});
    }
);
d.runAll();
app.post('/test', function (req, res) {
  res.send('Hello World!');
});

app.listen(3000, function () {
  console.log('Example app listening on port 3000!');
});
`

and error i get is : 
node_modules/durable/libjs/durableEngine.js:430
                    r.bindRuleset(handle, db.host, db.port, db.password);
Error: Could not create connection, error code: 301
    at Error (native)
    at Object.that.bind (/node_modules/durable/libjs/durableEngine.js:430:23)

**Comment by fmkone** [comment]
tried this as well:
app.use(durable.runAll) but didnt work either

**Comment by jruizgit** [comment]
Hi, in the case of node.js durable uses express to implement the REST API. You can reuse the express app as follows:

```javascript
var d = require('durable');
var m = d.m, s = d.s, c = d.c;
var timeout = d.timeout;

d.ruleset('test', {
    whenAll: [
        c.first = m.location.eq('US'),
        c.second = m.location.eq('CA')
    ],
    run: function(c) {
        console.log('test fraud detected ' + c.first.amount + ' ' + c.second.amount);
    }
},
function (host) {
    host.post('test', {id: 1, sid: 1, amount: 200, location: 'US'});
    host.post('test', {id: 2, sid: 1, amount: 200, location: 'CA'});
});

d.runAll(null, null, null, function(host, app) {
    app.get('/testhello', function (req, res) {
        res.send('Hello World!');
    });
    app.run();
});
```

Please note the path for the `get` function is /testhello, the path /test collides with the rule definition url.

**Comment by caimaoy** [comment]
```python
@flaskApp.route('/test/<sid>', methods=['POST'])
def post_event():
    message = json.loads(request.stream.read())
    message['sid'] = sid
    result = host.post('test', message)
    return make_response(json.dumps({'outcome': result}))
```
should be like below:

```python
@flaskApp.route('/test/<sid>', methods=['POST'])
def post_event(sid):
    message = json.loads(request.stream.read())
    message['sid'] = sid
    result = host.post('test', message)
    return make_response(json.dumps({'outcome': result}))
```
`def post_event():`  -> `def post_event(sid):`

**Comment by XXTest** [comment]
Does somebody has an implementation for "host = create_host()"?

**Comment by fmishin** [comment]
> Does somebody has an implementation for "host = create_host()"?
use instead:
host = get_host()

or 

instead of:
result = host.post('test', message)
call
result = post('test', message)
